back/boxtribute_server/models/crud.py

```diff
@@ -65,6 +80,170 @@ def create_box(
+        if distribution_event.state == DistributionEventState.Completed:
+            raise ModifyCompletedDistributionEvent(
+                desired_operation="move_box_to_distribution_event",
+                distribution_event_id=distribution_event.id,
+            )
+        box.location = distribution_event.distribution_spot_id
+        box.distribution_event = distribution_event_id
+        box.save()
+        return box
+
+
+def change_distribution_event_state(distribution_event_id, distribution_event_state):
+    distribution_event = DistributionEvent.get_by_id(distribution_event_id)
+
+    # Completed Events should not be mutable anymore
+    if distribution_event.state == DistributionEventState.Completed:
+        raise InvalidDistributionEventState(
+            expected_states=[
+                s
+                for s in DistributionEventState
+                if s != DistributionEventState.Completed
+            ],
+            actual_state=distribution_event_state,
+        )
+
+    distribution_event.state = distribution_event_state
+    distribution_event.save()
+    return distribution_event
+
+
+def add_packing_list_entry_to_distribution_event(
+    user_id,
+    distribution_event_id,
+    product_id,
+    size_id,
+    number_of_items,
+):
+    """
+    Add a packing list entry to a distribution event.
+    """
+    now = utcnow()
+
+    # Completed Events should not be mutable anymore
+    distribution_event = DistributionEvent.get_by_id(distribution_event_id)
+    if distribution_event.state == DistributionEventState.Completed:
+        raise ModifyCompletedDistributionEvent(
+            desired_operation="add_packing_list_entry",
+            distribution_event_id=distribution_event_id,
+        )
+
+    existing_packing_list_entry = PackingListEntry.get_or_none(
+        PackingListEntry.distribution_event == distribution_event_id,
+        PackingListEntry.product == product_id,
+        PackingListEntry.size == size_id,
+    )
+    if existing_packing_list_entry is not None:
+        if number_of_items > 0:
+            existing_packing_list_entry.number_of_items = number_of_items
+            existing_packing_list_entry.save()
+            return existing_packing_list_entry
+        else:
+            PackingListEntry.delete().where(
+                PackingListEntry.id == existing_packing_list_entry
+            ).execute()
+            return
+
+    else:
+        return PackingListEntry.create(
+            distribution_event=distribution_event_id,
+            product=product_id,
+            number_of_items=number_of_items,
+            size=size_id,
+            state=PackingListEntryState.NotStarted,
+            created_on=now,
+            created_by=user_id,
+            last_modified_on=now,
+            last_modified_by=user_id,
+        )
+
+
+def create_distribution_event(
+    user_id,
+    distribution_spot_id,
+    name,
+    planned_start_date_time,
```

Comment:
This is not required by the GraphQL schema but not optional here, probably should be `Datetime!` in the schema